File: Src/Layers/Layer.py
from Src.Utils.Values import layersNotToConsider
from Src.Utils.Shapes import get_shapes


def get_next_layer(outbound_nodes):
    next_layers = []
    if len(outbound_nodes) != 0:
        for out_node in outbound_nodes:
            nodes = out_node.outbound_layer

            if type(nodes) == list:
                for layer in nodes:
                    if layer.__class__.__name__ not in layersNotToConsider:
                        if layer not in next_layers:
                            next_layers.append(layer)
                    else:
                        next_layers_of_next = get_next_layer(layer.outbound_nodes)
                        next_layers = next_layers + next_layers_of_next
            else:
                if nodes.__class__.__name__ not in layersNotToConsider:
                    if nodes not in next_layers:
                        next_layers.append(nodes)
                else:
                    next_layers = get_next_layer(nodes.outbound_nodes)

    return next_layers


def get_prev_layer(inbound_nodes):
    prev_layers = []
    if len(inbound_nodes) != 0:
        for in_node in inbound_nodes:
            nodes = in_node.inbound_layers
            if type(nodes) == list:
                for layer in nodes:
                    if layer.__class__.__name__ not in layersNotToConsider:
                        if layer not in prev_layers:
                            prev_layers.append(layer)
                    else:
                        prev_layers_of_prev = get_prev_layer(layer.inbound_nodes)
                        prev_layers = prev_layers + prev_layers_of_prev
            else:
                if nodes.__class__.__name__ not in layersNotToConsider:
                    if nodes not in prev_layers:
                        prev_layers.append(nodes)
                else:
                    prev_layers = get_prev_layer(nodes.inbound_nodes)

    return prev_layers


class Layer:

    def __init__(self, shape, layer, model_inside_model, model_name):

        self.id = None
        self.selected = False
        self.lateral_space_between_layers = 10
        self.space_between_layers = 5
        self.center_position = [0, 0, 0]
        self.name = layer.name
        self.shape = shape
        self.original_model_layer = layer
        self.previous_layers = []
        self.next_layers = []
        self.activation = None
        self.type = layer.__class__.__name__
        self.input_shape = get_shapes(layer, True, True)
        self.output_shape = get_shapes(layer, False, True)
        self.invertedShape = layer.invertedShape
        self.computed_position = False
        self.layers = None
        self.model_inside_model = model_inside_model
        self.model_name = model_name

        self.previous_y_position = 0
        if hasattr(layer, "data_format"):
            self.data_format = layer.data_format
        try:
            self.activation = layer.activation.__name__
        except AttributeError:
            self.activation = None

    def set_id(self, identifier):
        self.id = identifier

    def set_x_position(self, x_position):
        self.center_position[0] = x_position + (self.shape[0]/2) + self.space_between_layers

    def set_y_position(self, y_position):
        self.center_position[2] = y_position

    def get_y_position(self):
        return self.center_position[2]

File: JavaScript_Src/Objects/Layer.js
class Layer {
    constructor(layer, copy = false) {
        this.id = layer.id;
        this.name = layer.name;
        this.type = layer.type;
        this.invertedShape = layer.invertedShape;
        this.activation = layer.activation;
        this.lastNegativeYPosition = 0;
        this.lastPositiveYPosition = 0;
        this.model_inside_model = layer.model_inside_model;
        this.model_name = layer.model_name;

        if (copy) {
            this.selected = layer.selected;
            this.centerPosition = layer.centerPosition;
            this.shape = [...layer.shape];
            this.prevLayers = layer.prevLayers;
            this.nextLayers = layer.nextLayers;
            this.activation = layer.activation;
            this.inputShape = layer.inputShape;
            this.outputShape = layer.outputShape;
            this.previousYPosition = layer.previousYPosition;
            this.batchNormalization = layer.batchNormalization;
            this.layers = layer.layers;

        } else {
            this.selected = false;
            this.centerPosition = [0, 0, 0];
            if (layer.type === "Dense" && layer.shape[0] === dynamicValues.minLength && layer.shape[1] === dynamicValues.minWidth)
                this.shape = [layer.shape[0], layer.shape[2], layer.shape[1]];
            else {
                this.shape = [...layer.shape];
            }
            this.prevLayers = layer.previous_layers;
            this.nextLayers = layer.next_layers;
            this.inputShape = layer.input_shape;
            this.outputShape = layer.output_shape;
            this.previousYPosition = 0;
            this.batchNormalization = layer.batch_normalization;
            this.layers = layer.layers;
            this.shouldBeDrawn = true;
            this.shouldBeBlock = false;
            this.isInsideBlock = false;
        }
    }

    setXPosition(xPosition) {
        this.centerPosition[0] = xPosition + (this.getShape()[0] / 2) + dynamicValues.spaceBetweenLayers;
    }

    setXPositionOfBeginBlockLayer(xPosition) {
        this.shape = [dynamicValues.blockSize, dynamicValues.blockSize, dynamicValues.blockSize];
        this.centerPosition[0] = xPosition + (this.getShape()[0] / 2) + dynamicValues.spaceBetweenLayers;
    }

    setXPositionInternalBlockLayer(xPosition) {
        this.shape = [dynamicValues.blockSize, dynamicValues.blockSize, dynamicValues.blockSize];
        this.centerPosition[0] = xPosition;
        this.isInsideBlock = true;
    }

    isLayerInsideBlock() {
        return this.isInsideBlock;
    }

    setYPosition(yPosition) {
        this.centerPosition[2] = yPosition;
    }

    getYPosition() {
        return this.centerPosition[2];
    }

    copy() {
        return new Layer(this, true);
    }

    getShape() {
        const typeOfDimension = document.getElementById('dimensionType');

        let layerShape = [];

        switch (typeOfDimension.value) {
            case "Exponential":
                layerShape = [...this.shape];
                if (layerShape[0] >= dynamicValues.minLength * 2)
                    layerShape[0] = layerShape[0] / Math.log(layerShape[0])
                if (layerShape[1] >= dynamicValues.minWidth * 2)
                    layerShape[1] = layerShape[1] / Math.log(layerShape[1])
                if (layerShape[2] >= dynamicValues.minWidth * 2)
                    layerShape[2] = layerShape[2] / Math.log(layerShape[2])
                break;
            case "Real":
                layerShape = [...this.shape];
                break;
            case "Linear":
                layerShape[0] = dynamicValues.minLength;
                layerShape[1] = dynamicValues.minWidth;
                layerShape[2] = dynamicValues.minWidth;
                break;
            default:
                break;
        }

        if (layerShape.length === 0) {
            layerShape[0] = dynamicValues.minLength;
            layerShape[1] = dynamicValues.minWidth;
            layerShape[2] = dynamicValues.minWidth;
        }

        layerShape[0] = Math.min(Math.max(layerShape[0], dynamicValues.minLength), dynamicValues.maxLength);
        layerShape[1] = Math.min(Math.max(layerShape[1], dynamicValues.minWidth), dynamicValues.maxWidth);
        layerShape[2] = Math.min(Math.max(layerShape[2], dynamicValues.minWidth), dynamicValues.maxWidth);

        return layerShape;
    }
}